core/trino-main/src/main/java/io/trino/cost/AggregationStatsRule.java

```diff
@@ -53,7 +54,7 @@ protected Optional<PlanNodeStatsEstimate> doCalculate(AggregationNode node, Stat
             return Optional.empty();
         }
 
-        if (node.getStep() != SINGLE) {
+        if (node.getStep() != SINGLE && node.getStep() != FINAL) {
```

Comment:
We could probably calculate the partial step output by `node count X driver count per node X empty grouping set count`.
Does it matter if we don't estimate for the partial step?

Reply:
there is no session available?

Reply:
Session is available but what properties could give you drivers count?

Reply:
That depends on the aggregation source. This can be either `TaskManagerConfig#getMinDrivers` if the node is in the source pipeline (e.g. after TableScan) or  `task_concurrency` otherwise. There is also `TaskCountEstimator` that estimates the number of tasks (nodes participating in the query)

Reply:
how to access this TaskManagerConfig? https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/Session.java

Reply:
PushPartialAggregationThroughExchange runs after all the CBO rules. So lack of estimates for partial aggregations is only a cosmetic problem.
Also, that was explicitly removed in https://github.com/prestodb/presto/pull/11595

Reply:
> PushPartialAggregationThroughExchange runs after all the CBO rules. So lack of estimates for partial aggregations is only a cosmetic problem.

but then we also don't need it for the `FINAL` step

Reply:
Yes, we don't, all we need to do here is to ensure that global aggregation estimate is populated for SINGLE step even when input stats are unknown.

Reply:
```
select orderkey, (select MIN(extendedprice) from hive.sf10.lineitem) as test from hive.sf10.lineitem;
```
Produce:

```
 Fragment 2 [SINGLE]
     CPU: 121.88ms, Scheduled: 143.44ms, Blocked 15.01s (Input: 10.98s, Output: 0.00ns), Input: 2526 rows (22.20kB); per task: avg.: 2526.00 std.dev.: 0.00, Output: 1 row (9B)
     Output layout: [min]
     Output partitioning: BROADCAST []
     Aggregate[type = FINAL]
     │   Layout: [min:double]
     │   Estimates: {rows: 1 (9B), cpu: ?, memory: 9B, network: 0B}
```

With this change, `Estimates` present 1 row as an output, which could change how output partitioning is chosen.

Reply:
You're looking at cost estimates of the final plan that is shown to the user. The CBO sees a different plan, where the  splitting of SINGLE aggregation has yet to happen.
You can try removing `&& node.getStep() != FINAL` change done here and see if the CBO still chooses BROADCAST join.

Reply:
Fix for show stats in separate commit